numpy/_core/src/umath/string_ufuncs.cpp

```diff
@@ -173,39 +116,38 @@ string_cmp(const character *str1, int elsize1, const character *str2, int elsize
 }
 
 
-template <typename character>
+template <ENCODING enc>
 static inline void
-string_add(character *str1, int elsize1, character *str2, int elsize2, character *out, int outsize)
+string_add(Buffer<enc> buf1, Buffer<enc> buf2, char *out)
 {
-    int len1 = get_length(str1, elsize1);
-    int len2 = get_length(str2, elsize2);
-
-    memcpy(out, str1, len1 * sizeof(character));
-    memcpy(out + len1, str2, len2 * sizeof(character));
-    memset(out + len1 + len2, 0, outsize - len1 - len2);
+    npy_int64 len1 = buf1.length();
+    npy_int64 len2 = buf2.length();
+    buf1.buffer_memcpy(out, len1);
+    buf2.buffer_memcpy_with_offset(out, len1, len2);
 }
 
 
-template <typename character>
+template <ENCODING enc>
 static inline npy_bool
-string_isalpha(const character *str, int elsize)
+string_isalpha(Buffer<enc> buf)
 {
-    int len = get_length(str, elsize);
+    npy_int64 len = buf.length();
```

Comment:
Maybe it should be called `num_codepoints()` instead of `length()` since the latter is ambiguous with the size in bytes.